### PR TITLE
JWT token auth bitches

### PR DESCRIPTION
**Pull request by asg0451**
switch from cookie-based sessions to using a JSON Web Token, which is a signed and encrypted json object with our information in it (and a TTL). Basically we auth with oauth as usual, and then compute and return this token to the client and forget about it. When we get a request, we validate the Authenticate header's contents which should be a valid token (signed by us, not expired, etc), and use its data to perform the request, i.e. what user is asking for stuff. 

this complicated signout slightly as we have no way of guaranteeing that the client will discard the token, so we have to have a db of logged_out tokens. this is implemented with memcached.

### CODE REVIEW

**Comment by asg0451** [comment]
[switch to stateless token auth in api](https://trello.com/c/dzMLs2sc/26-switch-to-stateless-token-auth-in-api)

**Comment by danielyaa5** [comment]
Come on put something in the descriptions....

**Comment by asg0451** [comment]
youre so needy dan gosh

**Comment by danielyaa5** [comment]
I wont bikeshed you, but Im a stickler for organization

Daniel Yakobian - djy18@case.edu
Case Western Reserve University - 2016

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,11 +1,44 @@
...
+
+  # [ {"data"=>"test"}, # payload
+  #   {"typ"=>"JWT", "alg"=>"whatever"} # header ]
+
+  def raw_jwt_token_str
+    # remove leading 'Bearer'
+    request.headers['Authorization']&.sub(/^Bearer /, '')
+  end
+
+  def decoded_token
```
Comment (danielyaa5) [inline app/controllers/application_controller.rb:14]: can we add a short description here

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,11 +1,44 @@
 class ApplicationController < ActionController::API
```
Comment (danielyaa5) [inline app/controllers/application_controller.rb:2]: A file overview here would be nice, is this a controller for the entire application?

Comment (asg0451) [reply app/controllers/application_controller.rb:2]: its the superclass of all controllers. really standard rails organization.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (asg0451) [reply app/controllers/application_controller.rb:14]: this is a super common rails pattern but ill add comments to this file for you guys

File: config/environments/test.rb [hunk 3]
```diff
@@ -39,4 +39,7 @@
 
   # Raises error for missing translations
   # config.action_view.raise_on_missing_translations = true
+
+  # hmac key for test
+  config.JWT_HMAC_SECRET = 'poop'
```
Comment (rxf0) [inline config/environments/test.rb:44]: Not secret enough

**Review by rxf0 — APPROVED** [review]

**Comment by danielyaa5** [comment]
Nice description? Why cant we gaurantee that user will discard the token?

**Comment by asg0451** [comment]
because a bad guy could just read it off a previous request in browser and continue to use it. it has to be invalidated somehow right?

**Comment by danielyaa5** [comment]
Not if we use https right?

Daniel Yakobian - djy18@case.edu
Case Western Reserve University - 2016

**Comment by asg0451** [comment]
what if the user was using a public computer? its just not safe without it.

**Review by danielyaa5 — APPROVED** [review]

**Pull request merged**